### PR TITLE
fix docs/errors for melt, measure, patterns, eval_with_cols: cols arg should not be provided by user

### LINKED ISSUE #5063: Odd behaviour of `melt()` when using the `cols` argument in `measure()`
`measure()` is a great enhancement to `melt()`  - thanks a lot for this!

However, while experimenting with the new `measure()` function when trying to find an answer for [this question on SO](https://stackoverflow.com/q/68233050/3817004) I came across an odd behaviour of `melt()` when the `cols` argument is used with `measure()`:

`#` [`Minimal reproducible example`](https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example)

```
df <- fread("Country    AXOX1991    AXOX1992    BXOX1991    BXOX1992    CXOX1991    CXOX1992
Afghanistan 1   2   3   4   5   6
USA 6   5   4   3   2   1")
```

This gives the expected result (note that the `AXO` columns intentionally have be left out of the reshape):

```
melt(df,  measure.vars = measure(var, year, 
                                 pattern = "([BC]XO)X(\\d{4})"))
```
```
       Country AXOX1991 AXOX1992 var year value
1: Afghanistan        1        2 BXO 1991     3
2:         USA        6        5 BXO 1991     4
3: Afghanistan        1        2 BXO 1992     4
4:         USA        6        5 BXO 1992     3
5: Afghanistan        1        2 CXO 1991     5
6:         USA        6        5 CXO 1991     2
7: Afghanistan        1        2 CXO 1992     6
8:         USA        6        5 CXO 1992     1
```

Now, when I include `cols` in the call to `measure()`

```
(cols <- names(df)[4:7])

[1] "BXOX1991" "BXOX1992" "CXOX1991" "CXOX1992"

melt(df,  measure.vars = measure(var, year, cols = cols,
                                 pattern = "([BC]XO)X(\\d{4})"))
```

I get an unexpected and obviously wrong result and a warning:

```
   BXOX1992 CXOX1991 CXOX1992 var year       value
1:        4        5        6 BXO 1991 Afghanistan
2:        3        2        1 BXO 1991         USA
3:        4        5        6 BXO 1992           1
4:        3        2        1 BXO 1992           6
5:        4        5        6 CXO 1991           2
6:        3        2        1 CXO 1991           5
7:        4        5        6 CXO 1992           3
8:        3        2        1 CXO 1992           4
```

> Warning message:
> In melt.data.table(df, measure.vars = measure(var, year,  :
>   'measure.vars' [Country, AXOX1991, AXOX1992, BXOX1991, ...] are not all of the same type. By order of hierarchy, the molten data value column will be of type 'character'. All measure variables not of type 'character' will be coerced too. Check DETAILS in ?melt.data.table for more on coercion.
> 

The documentation says on the `pattern` argument:

> Perl-compatible regex with capture groups to match to `cols`.

Perhaps, I have completely misunderstood the meaning of the `cols` argument. Unfortunately, I have found no example of the proper use of `cols` neither in the documentation of `measure()` nor in the reshape vignette.

--------------

Even stranger, when using a function to create the _character vector of column names_ in the call, I get an error message right away:

```
melt(df,  measure.vars = measure(var, year, cols = names(df)[4:7],
                                 pattern = "([BC]XO)X(\\d{4})"))
``` 

> Error in (function (fun.list, sep = "_", pattern, cols, multiple.keyword = "value.name",  : 
>   cols must be a character vector of column names

`#` `Output of sessionInfo()`

```
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19041)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252
[4] LC_NUMERIC=C                    LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.14.1

loaded via a namespace (and not attached):
[1] compiler_4.1.0 tools_4.1.0  
```

### PR DESCRIPTION
**Pull request by tdhock**
closes #5063 

Beta testing new measure() function revealed that users were confused about how to use cols arg.

Consensus from issue seems to be that docs for cols arg need to be clarified to indicate that users should NOT provide this argument.

### CODE REVIEW

**Comment by tdhock** [comment]
I don't think cols could be removed from measure()'s args.
Again, I based my design on patterns() which is defined as
```r
patterns = function(..., cols=character(0L)) {
```
melt() uses NSE via eval_with_cols() which automatically provides the cols arg, for user convenience, so the user does not have to manually specify cols.
For example consider the following code which was taken from ?patterns (note the user does not specify cols arg)
```r
     melt(DT, measure.vars = patterns("^x", "^y"))
```
So the code above is what the user types, but the code below is what gets evaluated, due to NSE/eval_with_cols(),
```r
     melt(DT, measure.vars = patterns("^x", "^y", cols=names(DT))
```
so actually when patterns (or measure/measurev) gets called in this way, even though the user did not provide the cols arg, the NSE/eval_with_cols() did provide cols arg, so cols is not missing. So that is the reason why Michael's idea won't work (he suggested making !missing(cols) an error).

Sorry about all the confusion / hassle.

**Comment by tdhock** [comment]
So rather than adding logic to measure/measurev to try to detect the missing cols arg (which would not be possible, given that it is filled in via NSE), I implemented an error message in `eval_with_cols`, if the user provides the `cols` argument. I also clarified the documentation of the `cols` arg and added an example of usage with .SDcols in ?patterns man page.

**Comment by tdhock** [comment]
The new behavior is the error message below, which was previously not present, but now clearly indicates that the user should remove the cols argument in their code. 
Previously, other less informative errors could result (confusing for user), like in #5063, if user provided cols which are not equal to actual column names. 
```r
> i1=data.table(iris)[1]
> melt(i1, measure=patterns("Sepal"))
   Petal.Length Petal.Width Species     variable value
          <num>       <num>  <fctr>       <fctr> <num>
1:          1.4         0.2  setosa Sepal.Length   5.1
2:          1.4         0.2  setosa  Sepal.Width   3.5
> melt(i1, measure=patterns("Sepal", cols=names(iris)))
Erreur : user should not provide cols argument to patterns
```

**Comment by tdhock** [comment]
hi @jangorecki @MichaelChirico I would like to include this fix in next release, and I believe that it is ready to merge into master, so could you please review and/or merge if you agree that it is ready? Thanks!

**Comment by MichaelChirico** [comment]
> next release

Just to clarify, the current milestone (1.16.0) is correct right? If so I will do review on PRs in the target milestone "regularly", so no need to ping.

File: R/utils.R [hunk 1]
```diff
@@ -132,8 +132,12 @@ eval_with_cols = function(orig_call, all_cols) {
   })
   if (!is.primitive(fun)) {
     named_call = match.call(fun, orig_call)
-    if ("cols" %in% names(formals(fun)) && !"cols" %in% names(named_call)) {
-      named_call[["cols"]] = all_cols
+    if ("cols" %in% names(formals(fun))) {
+      if ("cols" %in% names(named_call)) {
+        stopf("user should not provide cols argument to %s", paste(fun_uneval))
```
Comment (MichaelChirico) [inline R/utils.R:137]: Can `length(fun_uneval)>1`? It looks like it's coming from `substitute(call)[[1]]`. That can be `length>1` in namespaced calls:

```r
length(substitute(stats::lm(x))[[1]])
[1] 3
```

Is that possibly the input here? If that case was not considered previously we might file a follow-up bug to handle this.

Relevant to the current PR because that `paste()` call might not WAI.

In any case, I would prefer `as.character()` to make the intention explicit.

Comment (MichaelChirico) [reply R/utils.R:137]: Second observation: this error will not be very actionable. As a "clueless" user I may get this error and worry "What did I do wrong? What does this mean?" More helpful might be to suggest an alternative, or point to a helpful doc page.

Comment (MichaelChirico) [reply R/utils.R:137]: > Is that possibly the input here?

Answering my own question based on the tests: Looks like no. But we should add `stopifnot(length(fun_uneval) == 1)` -- most importantly we might eventually export placeholders for package authors (like we do with `` `:=` ``) as masks for `R CMD check` issues, and then authors might be tempted to use `data.table::measure` leading to the above issue.

File: inst/tests/tests.Rraw [hunk 2]
```diff
@@ -17240,12 +17241,13 @@ test(2182.6, melt(DT.wide, measure.vars=list(b=c("b1","b2"))), data.table(a2=2,
...
 measurev = list("foo", "bar")#measurev below should not use this since it is not a function.
 test(2183.00002, melt(DTid, measure.vars=measurev(list(value.name=NULL, num=as.complex), pattern="([ab])([12])")), error="Type 'complex' is not supported for joining/merging")
 test(2183.00004, melt(DTid, measure.vars=measurev(list(value.name=NULL, istr=NULL), pattern="([ab])([12])"))[order(b)], data.table(id=1, istr=paste(c(1,2)), a=c(NA, 2), b=c(1,2)))
 test(2183.00005, melt(DTid, measure.vars=measurev(list(column=NULL, istr=NULL), pattern="([ab])([12])", multiple.keyword="column"))[order(b)], data.table(id=1, istr=paste(c(1,2)), a=c(NA, 2), b=c(1,2)))#same computation but different multiple.keyword
 iris.dt = data.table(datasets::iris)
-test(2183.00020, melt(iris.dt, measure.vars=measurev(value.name, dim, sep=".", pattern="foo")), error="both sep and pattern arguments used; must use either sep or pattern (not both)")
+test(2183.00020, melt(iris.dt, measure.vars=measurev(list(value.name=NULL, dim=NULL), sep=".", pattern="foo")), error="both sep and pattern arguments used; must use either sep or pattern (not both)")
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:17250]: test change appears unrelated to the rest of the PR?

File: inst/tests/tests.Rraw [hunk 3]
```diff
@@ -17334,9 +17337,10 @@ test(2183.74, melt(DTid, measure.vars=measure(letter, number, multiple.keyword=a
...
 test(2183.76, melt(DTid, measure.vars=measure(letter, number, multiple.keyword="", pattern="([ab])([12])")), error="multiple.keyword must be a character string with nchar>0")
 test(2183.77, melt(DTid, measure.vars=measure(letter, cols, pattern="([ab])([12])")), error="group names specified in ... conflict with measure argument names; please fix by changing group names: [cols]")
-test(2183.78, melt(DTid, measure.vars=measure(letter, cols=as.integer, pattern="([ab])([12])")), error="cols must be a character vector of column names")
+test(2183.78, melt(DTid, measure.vars=measure(letter, cols=as.integer, pattern="([ab])([12])")), error="user should not provide cols argument to measure")
 test(2183.79, melt(DTid, measure.vars=measure(letter, number, pattern=as.integer)), error="pattern must be character string")
 test(2183.80, melt(DTid, measure.vars=measure(letter, number, sep=as.integer)), error="sep must be character string")
+test(2183.81, data.table:::measure(cols=as.integer), error="cols must be a character vector of column names")
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:17343]: Prefer a local mask for unexported functions, see the top of this file (sorry, can't permalink easily on this massive file, but it's right at the top)

File: R/utils.R [hunk 4]
```diff
@@ -132,8 +132,12 @@ eval_with_cols = function(orig_call, all_cols) {
   })
   if (!is.primitive(fun)) {
     named_call = match.call(fun, orig_call)
-    if ("cols" %in% names(formals(fun)) && !"cols" %in% names(named_call)) {
-      named_call[["cols"]] = all_cols
+    if ("cols" %in% names(formals(fun))) {
+      if ("cols" %in% names(named_call)) {
```
Comment (MichaelChirico) [inline R/utils.R:136]: Unnecessary nesting here?

```suggestion
    if ("cols" %in% names(formals(fun)) && "cols" %in% names(named_call)) {
```

Comment (MichaelChirico) [reply R/utils.R:136]: NVM, I'm wrong, the nesting is necessary because of the `else`.

File: inst/tests/tests.Rraw [hunk 5]
```diff
@@ -12305,7 +12305,7 @@ DTout = data.table(
   value1 = 1:10,
   value2 = c("a", "b", "c", "d", "e", "f", "g", "h", "i", "j")
 )
-test(1866.6, melt(DT, measure.vars = patterns("^x", "^y", cols=names(DT))), DTout)
+test(1866.6, melt(DT, measure.vars = patterns("^x", "^y")), DTout)
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:12308]: Isn't this a potentially breaking change vs. 1.15.0?

File: man/measure.Rd [hunk 6]
```diff
@@ -38,7 +38,10 @@ measurev(fun.list, sep, pattern, cols, multiple.keyword="value.name",
     are considered measure variables.}
   \item{pattern}{Perl-compatible regex with capture groups to match to
     \code{cols}. Columns that match the regex are considered measure variables.}
-  \item{cols}{A character vector of column names.}
+  \item{cols}{A character vector that is automatically provided,
```
Comment (MichaelChirico) [inline man/measure.Rd:41]: I'm still not sure I understand (sorry, I'm not caught up yet on the new functionality of `melt()`). Can we simply remove this argument then? Or it's needed for internal calls to `melt()`? If that's the case, let's make that more evident?

```
\item{cols}{Do not use. This will be populated automatically with all of the column names of the data.table input to melt. To specify the columns to be measured, use the \code{pattern} argument.}
```

File: man/patterns.Rd [hunk 7]
```diff
@@ -16,18 +16,25 @@ patterns(\dots, cols=character(0))
...
+    user should not provide the \code{cols} argument, which is
+    automatically set to all of the column names of the input data table.
+  }
 }
 \seealso{
-  \code{\link{melt}},
-  \url{https://github.com/Rdatatable/data.table/wiki/Getting-started}
+  \code{\link{melt}}, \code{\link{.SD}},
+  \url{https://github.com/Rdatatable/data.table/wiki/Getting-started},
+  \url{https://cloud.r-project.org/web/packages/data.table/vignettes/datatable-reshape.html},
```
Comment (MichaelChirico) [inline man/patterns.Rd:30]: nit: possibly preferable to use `vignette("datatable-reshape")`? that way users reading this page from an airgapped machine can still access the resource.

Or perhaps, whatever the Rd equivalent of markdown:

```markdown
[`vignette("datatable-reshape")`](https://cloud.r-project.org/web/packages/data.table/vignettes/datatable-reshape.html)
```

would be to get the best of both worlds (link the HTML version where links are available, but show the code to access from anywhere)?

File: inst/tests/tests.Rraw (hunk 5 above)
Comment (tdhock) [reply inst/tests/tests.Rraw:12308]: yes, but I think it is appropriate
- there was only one test case
- specifying cols like this is redundant
- there is no other valid value for cols (other than names(DT))

File: man/measure.Rd (hunk 6 above)
Comment (tdhock) [reply man/measure.Rd:41]: we can not simply remove this argument, because it needs to be provided by eval_with_cols (for both measure and patterns).
I can update the docs to say "do not use" but based on your prior comments about maybe one day exporting this function, that may be confusing. What we really want is: users should not provide cols argument to measure or patterns, when they are used in the context of melt or [.data.table.

File: inst/tests/tests.Rraw (hunk 3 above)
Comment (jangorecki) [reply inst/tests/tests.Rraw:17343]: this will not get tested during `cc()` afaict

File: inst/tests/tests.Rraw (hunk 5 above)
Comment (jangorecki) [reply inst/tests/tests.Rraw:12308]: we need to be sure it is well mentioned in NEWS

File: man/measure.Rd (hunk 6 above)
Comment (jangorecki) [reply man/measure.Rd:41]: couldn't we have 2 functions then, one internal and one meant to be used by users?

Comment (tdhock) [reply man/measure.Rd:41]: I don't think so. I think we just need to document the NSE better, I will give it a try.

File: inst/tests/tests.Rraw (hunk 3 above)
Comment (tdhock) [reply inst/tests/tests.Rraw:17343]: I see the block of local definitions for un-exported functions, but I don't think it will work in this case, because there are a few tests, for example 2183.01, which require defining another object named measure (would over-write) so I think it is better to just explicitly use data.table:::measure, is that ok?

File: R/utils.R [hunk 8]
```diff
@@ -134,7 +135,7 @@ eval_with_cols = function(orig_call, all_cols) {
     named_call = match.call(fun, orig_call)
     if ("cols" %in% names(formals(fun))) {
       if ("cols" %in% names(named_call)) {
-        stopf("user should not provide cols argument to %s", paste(fun_uneval))
+        stopf("user should not provide cols argument to %s, when specifying the columns for melt or .SDcols; in this context, non-standard evaluation is used internally to set cols to all data table column names, so please fix by removing cols argument", paste(fun_uneval))
```
Comment (tdhock) [inline R/utils.R:138]: here is my attempt at more complete error message

File: man/measure.Rd [hunk 9]
```diff
@@ -38,10 +37,15 @@ measurev(fun.list, sep, pattern, cols, multiple.keyword="value.name",
...
-    to specify the columns to be measured, use the pattern argument.}
+  \item{cols}{
+    A character vector of column names.
+    When used in the context of
+    \code{melt(measure.vars=measure())}, the
+    user should not provide the \code{cols} argument, which is
+    automatically set to all of the column names of the input data
+    table, using non-standard evaluation.
+    To specify the columns to be measured,
+    use either \code{sep} or \code{pattern} argument.}
```
Comment (tdhock) [inline man/measure.Rd:48]: here is my attempt to clarify NSE in the measure() docs, is this OK? similar wording is used in patterns().

File: man/melt.data.table.Rd [hunk 10]
```diff
@@ -21,16 +21,16 @@ multiple columns simultaneously.
...
-\item{measure.vars}{Measure variables for \code{melt}ing. Can be missing, vector, list, or pattern-based.
+\item{measure.vars}{Measure variables for \code{melt}ing. Can be missing, vector, list, or function.
 
   \itemize{
     \item{ When missing, \code{measure.vars} will become all columns outside \code{id.vars}. }
     \item{ Vector can be \code{integer} (implying column numbers) or \code{character} (column names). }
     \item{ \code{list} is a generalization of the vector version -- each element of the list (which should be \code{integer} or \code{character} as above) will become a \code{melt}ed column. }
-    \item{ Pattern-based column matching can be achieved with the regular expression-based \code{\link{patterns}} syntax; multiple patterns will produce multiple columns. }
+    \item{ Pattern-based column matching can be achieved by using the result of \code{\link{patterns}} or \code{\link{measure}} functions. More generally, it is possible to use any function that returns a character vector or list of measure variables, and if this function has an argument named \code{cols}, then that argument will be set to \code{names(data)} by non-standard evaluation. }
```
Comment (tdhock) [inline man/melt.data.table.Rd:30]: here is documentation of the non-standard evaluation on ?melt

File: man/data.table.Rd [hunk 11]
```diff
@@ -152,6 +152,10 @@ data.table(\dots, keep.rownames=FALSE, check.names=FALSE, key=NULL, stringsAsFac
...
 
     Finally, you can filter columns to include in \code{.SD} based on their \emph{names} according to regular expressions via \code{.SDcols=patterns(regex1, regex2, ...)}. The included columns will be the \emph{intersection} of the columns identified by each pattern; pattern unions can easily be specified with \code{|} in a regex. You can filter columns on \code{values}  by passing a function, e.g. \code{.SDcols=\link{is.numeric}}. You can also invert a pattern as usual with \code{.SDcols=!patterns(...)} or \code{.SDcols=!is.numeric}.
+
+    Note that if there is a user-defined function named \code{patterns}, it will be used instead of the usual patterns, and it should return a list of column names to intersect, and if that function has an argument named \code{cols}, then that argument will be set to \code{names(x)} by non-standard evaluation.
+
+    In addition, \code{.SDcols=fun_not_named_patterns()} is possible, and that function should return a character vector of column names, and non-standard evaluation is not used, even if there is an argument named \code{cols}.
```
Comment (tdhock) [inline man/data.table.Rd:158]: This is documentation of the current behavior of .SDcols and patterns, which I think is pretty confusing.
Currently, the function name patterns is a special case, for which we use NSE to fill in the cols argument.
But if the user does .SDcols=any_name_other_than_patterns(), then NSE is not used to fill in the cols argument of that function, which is inconsistent and potentially confusing. 
For consistency, simplicity, and to make the documentation easier to understand, I would suggest modifying the code so that patterns() is no longer a special case, and NSE is always used (with any function name).
I'm not sure that would be a "breaking change" since this is currently un-documented and un-tested behavior.
Any comments/opinions/feedback?

File: inst/tests/tests.Rraw (hunk 3 above)
Comment (jangorecki) [reply inst/tests/tests.Rraw]: Not fine. data.table::: can be used only at the too of the test script

Comment (tdhock) [reply inst/tests/tests.Rraw]: ok, I fixed it!

File: inst/tests/tests.Rraw (hunk 2 above)
Comment (tdhock) [reply inst/tests/tests.Rraw]: you are right. should I exclude/revert? or is it ok to keep? It is somewhat related to the theme of the PR ... (measure...)

File: R/utils.R (hunk 1 above)
Comment (tdhock) [reply R/utils.R]: ok i added the stopifnot and I changed paste to as.character.

File: R/utils.R [hunk 12]
```diff
@@ -120,6 +120,7 @@ brackify = function(x, quote=FALSE) {
 eval_with_cols = function(orig_call, all_cols) {
   parent = parent.frame(2L)
   fun_uneval = orig_call[[1L]]
+  stopifnot(length(fun_uneval) == 1)
```
Comment (MichaelChirico) [inline R/utils.R:123]: IINM we need a regression test here?

File: R/utils.R [hunk 13]
```diff
@@ -132,8 +133,12 @@ eval_with_cols = function(orig_call, all_cols) {
...
   if (!is.primitive(fun)) {
     named_call = match.call(fun, orig_call)
-    if ("cols" %in% names(formals(fun)) && !"cols" %in% names(named_call)) {
-      named_call[["cols"]] = all_cols
+    if ("cols" %in% names(formals(fun))) {
+      if ("cols" %in% names(named_call)) {
+        stopf("user should not provide cols argument to %s, when specifying the columns for melt or .SDcols; in this context, non-standard evaluation is used internally to set cols to all data table column names, so please fix by removing cols argument", as.character(fun_uneval))
+      } else {
+        named_call[["cols"]] = all_cols
+      }
```
Comment (MichaelChirico) [inline R/utils.R:141]: I call the `else` branch here "unnecessary nesting" -- the `stop()` branch will always exit, so we can reduce the indentation of the non`-stop()` branch. (NB: this is something caught by `lintr::unnecessary_nesting_linter()`, we can address this generally in the repo if we so choose.)

```suggestion
      }
      named_call[["cols"]] = all_cols
```

File: inst/tests/tests.Rraw [hunk 14]
```diff
@@ -13955,6 +13956,18 @@ DT = data.table(
   V9 = c(0.2, -0.1, 1.2, -0.5, 1.4, 1, 0.2, 0.7, 0.4, 1.6),
   V10 = c(0.8, 0.7, -1.2, -0.9, -0.6, 0.4, -2.3, 2.2, 0.5, -1.4)
 )
+test(1971.01, DT[ , lapply(.SD, sum), .SDcols = patterns('^V', cols="V7")], error="user should not provide cols argument to patterns, when specifying the columns for melt or .SDcols; in this context, non-standard evaluation is used internally to set cols to all data table column names, so please fix by removing cols argument")
+not_patterns <- function()c("V4","V8")
+test(1971.02, DT[ , lapply(.SD, sum), .SDcols = not_patterns()], data.table(V4=3.4, V8=-0.4))
+not_patterns <- function(cols)cols
+test(1971.03, DT[ , lapply(.SD, sum), .SDcols = not_patterns()], error='argument "cols" is missing, with no default')
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:13963]: this is just a test of base R behavior, right? if so, we should remove it -- we are just setting ourselves up to be broken if base R changes this error message in any way, and IIRC is susceptible to failing in non-English test environments as well.

if you think this test captures something from data.table, please follow the precedent of `base_messages` (circa L159) for how to test against messages "owned" by R itself

File: inst/tests/tests.Rraw [hunk 15]
```diff
@@ -13955,6 +13956,18 @@ DT = data.table(
...
+test(1971.04, DT[ , lapply(.SD, sum), .SDcols = not_patterns(c("V4","V8"))], data.table(V4=3.4, V8=-0.4))
+patterns <- function(cols)list(cols[c(6,10)])#with cols arg
```
Comment (MichaelChirico) [inline inst/tests/tests.Rraw:13965]: ooh this is funky. this feels like something users should not rely on working... I would probably drop this test. WDYT?

Comment (tdhock) [reply inst/tests/tests.Rraw:13965]: yes I agree this is funky. Did you see my comment about this above? https://github.com/Rdatatable/data.table/pull/5115#pullrequestreview-1904158739 So to move forward I would suggest either
* keep current behavior, NSE in .SDcols only used with a function named patterns(), with user-defined patterns() un-documented and un-tested.
* modify current behavior so that NSE is used in .SDcols with any user-defined function (whether named patterns or not), and add new docs/tests. I believe this option should be preferable for consistency of how functions are evaluated in the context of .SDcols, and how eval_with_cols is used here and in melt.

@MichaelChirico can you please tell me which option you would prefer? After there is consensus, I will update the PR.

Comment (MichaelChirico) [reply inst/tests/tests.Rraw:13965]: I lean towards the former. We really are trying to minimize our use of NSE across the API, adding new places where it works outside `i=`, `j=`, `by=` is not recommended IMO. See e.g. #6035 where even very basic NSE is rejected.

File: inst/tests/tests.Rraw (hunk 14 above)
Comment (tdhock) [reply inst/tests/tests.Rraw]: ok I removed these tests

File: inst/tests/tests.Rraw (hunk 15 above)
Comment (tdhock) [reply inst/tests/tests.Rraw]: ok I removed the docs and tests which mention user-defined patterns

File: R/utils.R (hunk 12 above)
Comment (tdhock) [reply R/utils.R:123]: I'm not sure I understand, can you please clarify/suggest a test?

Comment (MichaelChirico) [reply R/utils.R:123]: Hmm, actually, in the current PR, this change looks out of place, is it still needed?

Comment (tdhock) [reply R/utils.R:123]: I made this change to address your comment https://github.com/Rdatatable/data.table/pull/5115#discussion_r1493949195 and actually I'm not sure why it is necessary, so you tell me, haha.

Comment (MichaelChirico) [inline R/utils.R:123]:
```suggestion
```

Comment (MichaelChirico) [reply R/utils.R]: :upside_down_face: sorry about that.

I think I was originally worried about usage like `data.table::patterns()` instead of the usual (unqualified) `patterns()`, but I don't think it actually matters here, reading again. We can wait for an actual bug report if there is such case; reverted for now.

File: R/utils.R [hunk 16]
```diff
@@ -132,7 +132,10 @@ eval_with_cols = function(orig_call, all_cols) {
   })
   if (!is.primitive(fun)) {
     named_call = match.call(fun, orig_call)
-    if ("cols" %in% names(formals(fun)) && !"cols" %in% names(named_call)) {
+    if ("cols" %in% names(formals(fun))) {
+      if ("cols" %in% names(named_call)) {
+        stopf("user should not provide cols argument to %s, when specifying the columns for melt or .SDcols; in this context, non-standard evaluation is used internally to set cols to all data table column names, so please fix by removing cols argument", as.character(fun_uneval))
```
Comment (MichaelChirico) [inline R/utils.R:137]: I think this is a much more serious breaking change than #5247 -- just glancing at GitHub shows lots of people using `cols=`:

https://github.com/search?q=lang%3AR%20%2Fpatterns%5B(%5D%5B%5E)%5D*cols%5Cs*%3D%2F&type=code

Even though I almost exclusively see it done with `cols=names(x)`, such usage is now an error. We need to remove the `cols=` argument more gradually.

Comment (tdhock) [reply R/utils.R:137]: ok change to warning?

Comment (MichaelChirico) [reply R/utils.R:137]: Let's map out a deprecation plan, e.g.

1.16.0: `warning()` if `cols` is used _for a subset of columns_, `message()` if it's used _for the full set of columns_. The difference because the latter case should simply drop `cols=`, the former case can actually get different behavior.
1.17.0: Upgrade to error to provide `cols=`
1.18.0: Remove code related to `cols=`

We can consider an option to allow the old behavior to keep working as well, in which case we probably want to lengthen the deprecation cycle by one more release.

Comment (tdhock) [reply R/utils.R:137]: another option is that we could continue to support user-provided `cols` and modify `measure()` to support that (would be possible I think if we make measure return character column names instead of integer column indices). However I think it would be better to do this warning/error, and encourage users to specify the right regex/pattern arg to patterns/measure. Thoughts?

Comment (MichaelChirico) [reply R/utils.R:137]: I don't have a strong prior here, though it does sound workable; do you think that would be useful and worth maintaining?

One thought is that it's sometimes clunky to jam everything into a pattern-based approach, e.g. it's not uncommon to have a vector of column names handy already, in which case just plugging it into `cols=` can be convenient as compared to turning that into a regex (esp if the column names might have special regex characters like `(` `{` `.` etc). That said we haven't seen a request to this end.
